include/aspect/mesh_deformation/interface.h

```diff
@@ -0,0 +1,398 @@
+         const std::string &description,
+         void (*declare_parameters_function) (ParameterHandler &),
+         Interface<dim> *(*factory_function) ());
+
+
+        /**
+         * Return a list of names of all mesh deformation models currently
+         * used in the computation, as specified in the input file.
+         */
+        const std::vector<std::string> &
+        get_active_mesh_deformation_names () const;
+
+        /**
+         * Return a list of pointers to all mesh deformation models
+         * currently used in the computation, as specified in the input file.
+         */
+        const std::vector<std_cxx11::shared_ptr<Interface<dim> > > &
+        get_active_mesh_deformation_models () const;
+
+        /**
+         * Go through the list of all mesh deformation models that have been selected in
+         * the input file (and are consequently currently active) and see if one
+         * of them has the desired type specified by the template argument. If so,
+         * return a pointer to it. If no mesh deformation model is active
+         * that matches the given type, return a NULL pointer.
+         */
+        template <typename MeshDeformationType>
+        MeshDeformationType *
+        find_mesh_deformation_model () const;
+
+        /**
+         * For the current plugin subsystem, write a connection graph of all of the
+         * plugins we know about, in the format that the
+         * programs dot and neato understand. This allows for a visualization of
+         * how all of the plugins that ASPECT knows about are interconnected, and
+         * connect to other parts of the ASPECT code.
+         *
+         * @param output_stream The stream to write the output to.
+         */
+        static
+        void
+        write_plugin_graph (std::ostream &output_stream);
+
+        /**
+         * Exception.
+         */
+        DeclException1 (ExcMeshDeformationNameNotFound,
+                        std::string,
+                        << "Could not find entry <"
+                        << arg1
+                        << "> among the names of registered mesh deformation objects.");
+
+      private:
+        /**
+         * A list of mesh deformation objects that have been requested in the
+         * parameter file.
+         */
+        std::vector<std_cxx11::shared_ptr<Interface<dim> > > mesh_deformation_objects;
+
+        /**
+         * A list of names of mesh deformation objects that have been requested
+         * in the parameter file.
+         */
+        std::vector<std::string> model_names;
+
+        /**
+         * Set the boundary conditions for the solution of the elliptic
+         * problem, which computes the displacements of the internal
+         * vertices so that the mesh does not become too distorted due to
+         * motion of the free surface.  Velocities of vertices on the free
+         * surface are set to be the normal of the Stokes velocity solution
```

Comment:
Maybe we should update the comments to include the "vertical" projection and the `Additional tangential mesh velocity boundary indicators`.